core/src/HIP/Kokkos_HIP_Instance.cpp

```diff
@@ -371,7 +371,17 @@ void HIPInternal::release_team_scratch_space(int scratch_pool_id) {
 //----------------------------------------------------------------------------
 
 void HIPInternal::finalize() {
+  // First, lock the shared resource locking helper.
+  // Then, fence the stream and check if it was involved in the last constant
+  // memory launch.
+  // Locking is required to avoid a race condition, i.e. it prevents another
+  // thread from launching another kernel in-between the fence
+  // and the 'check_if_involved'.
+  auto lock = HIPInternal::constantMemReusable[m_hipDev].lock();
   this->fence("Kokkos::HIPInternal::finalize: fence on finalization");
+  HIPInternal::constantMemReusable[m_hipDev].check_if_involved(std::move(lock),
+                                                               m_stream);
```

Comment:
`check_if_involved` doesn't suggest that the function will unlock the lock. Maybe,
```suggestion
  HIPInternal::constantMemReusable[m_hipDev].check_if_involved_and_unlock(std::move(lock),
                                                               m_stream);
```

Reply:
I might be wrong but it's a unique lock, so the function explicitly takes ownership of the lock, and does not return it to you. So I thought it was quite clear already that it would be released ?

```c++
#include <iostream>
#include <memory>
#include <mutex>

std::mutex mutex;

struct CustomLock
{
  void check(std::unique_lock<std::mutex> lock) const {
    std::cout << "Taking ownership of lock " << lock.mutex() << std::endl;
  }
};

int main()
{
  std::unique_lock<std::mutex> lock(mutex);

  std::cout << "Lock is mine: " << std::boolalpha << lock.owns_lock() << std::endl;

  // CustomLock{}.check(lock); // does not compile
  CustomLock{}.check(std::move(lock));
}
```

Reply:
What I mean is that it's hard to tell on first glance that the lock is moved implies ownership and thus the function hopefully unlocks the mutex. When I see `lock` I expect to also see `unlock` somewhere.

Reply:
OK, `check_if_involved_and_unlock` adopted :heavy_check_mark: